include/s3select.h

```diff
@@ -1,5 +1,6 @@
 #ifndef __S3SELECT__
 #define __S3SELECT__
+#define BOOST_SPIRIT_THREADSAFE
```

Comment:
is this part still necessary now that we're passing the s3select pointer into our actions?

Reply:
its their recommendation, i prefer not to challenge that (it should be there on initial commit, my mistake)

**_On the other hand, if a grammar is intended to be used in multithreaded code, we should then define BOOST_SPIRIT_THREADSAFE before including any spirit header files. In this case it will also be required to link against Boost.Threads_**

Reply:
ok. when i read that, i assumed it was talking about using an instance of `bsc::grammar<s3select>` across threads, which we aren't doing. i'd really like to avoid the cost of locks/atomics if we can

do you have a multi-threaded test case that we could build and run with thread sanitizer? passing that test should give us the confidence to leave it off - and failing would tell us exactly where/why we need it on

you can enable tsan in cmake like this:
```diff
diff --git a/test/CMakeLists.txt b/test/CMakeLists.txt
index 0d5dcfc..aae49f6 100644
--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -1,6 +1,7 @@
 add_executable(s3select_test s3select_test.cpp)
 target_include_directories(s3select_test PUBLIC ../include)
-target_link_libraries(s3select_test gtest gtest_main boost_date_time boost_thread boost_system )
+target_compile_options(s3select_test PUBLIC "-fsanitize=thread")
+target_link_libraries(s3select_test gtest gtest_main boost_date_time boost_thread boost_system "-fsanitize=thread")
 
 include(GoogleTest)
 gtest_discover_tests(s3select_test)
```

Reply:
- using thread-sanitizer , it detects data-race , its upon calling the parse(boost).
- the BOOST_SPIRIT_THREADSAFE has no impact (so far, per my tests)
- all queries execution per all threads are OK(no failures in teuthology, or any other tests)
- need to check that issue further(it may take some time), so i suggest to fix that issue in another PR.

Reply:
sounds good thanks for testing!